drv/lpc55-spi-server/src/main.rs

```diff
@@ -117,172 +79,64 @@ fn main() -> ! {
-                    // bogus data for now
-                    datref.spi.drain();
-                    datref.spi.enable_tx();
-                    datref.spi.enable_rx();
-
-                    ringbuf_entry!(Trace::Start(Op::Read));
-                    Ok(())
-                }
-                Op::Exchange => {
-                    let ((), caller) =
-                        msg.fixed_with_leases(2).ok_or(ResponseCode::BadArg)?;
-
-                    if datref.dat.is_some() {
-                        return Err(ResponseCode::Busy);
-                    }
-
-                    let borrow_send = caller.borrow(0);
-                    let send_info =
-                        borrow_send.info().ok_or(ResponseCode::BadArg)?;
-                    if !send_info.attributes.contains(LeaseAttributes::READ) {
-                        return Err(ResponseCode::BadArg);
-                    }
-
-                    // Keep a minimum size to go with the FIFO settings
-                    if send_info.len < 8 {
-                        return Err(ResponseCode::BadArg);
-                    }
-
-                    let borrow_recv = caller.borrow(1);
-                    let recv_info =
-                        borrow_recv.info().ok_or(ResponseCode::BadArg)?;
-                    if !recv_info.attributes.contains(LeaseAttributes::WRITE) {
-                        return Err(ResponseCode::BadArg);
-                    }
-
-                    if recv_info.len != send_info.len {
-                        return Err(ResponseCode::BadArg);
-                    }
-
-                    datref.dat = Some(Transmit {
-                        task: caller,
-                        rpos: 0,
-                        rlease_num: 0,
-                        wpos: 0,
-                        wlease_num: 1,
-                        len: recv_info.len,
-                        op: Op::Exchange,
-                    });
-
-                    datref.spi.drain();
-                    datref.spi.enable_tx();
-                    datref.spi.enable_rx();
-
-                    ringbuf_entry!(Trace::Start(Op::Exchange));
-                    Ok(())
-                }
-            },
-        );
+        let _m = match sys_recv_closed(&mut [], 1, TaskId::KERNEL) {
+            Ok(m) => m,
+            Err(_) => panic!(),
+        };
+
+        ringbuf_entry!(Trace::IRQ);
+
+        while spi.can_tx() && !tx_done {
+            let b = tx[tx_cnt];
+            ringbuf_entry!(Trace::Tx(b));
+            spi.send_u8(tx[tx_cnt]);
+            tx_cnt += 1;
+            if tx_cnt == tx.len() {
+                tx_done = true;
+            }
+        }
+
+        while spi.has_byte() && !rx_done {
+            let b = spi.read_u8();
+            ringbuf_entry!(Trace::Rx(b));
+            rx[rx_cnt] = b;
+            rx_cnt += 1;
+            if rx_cnt == rx.len() {
+                rx_done = true;
+            }
+        }
+
+        if tx_done && rx_done {
```

Comment:
What happens if the loops bail early and you never get `tx_done` or `rx_done`? Is that a safe condition?

Reply:
this is designed to be used in very limited circumstances right now, mostly for hardware signal testing. Even so, I'm not sure I understand what you mean by "safe".  If `tx_done` and `rx_done` are not set this will loop to wait for the next interrupt that either the TX or RX FIFOs are ready. It's fine to wait indefinitely.

Reply:
Thanks!

Reply:
Thanks. I just meant was there any special condition that would need to be taken the next time around the loop. Looks like the answer is no, which clarifies my understanding.